Fix/refine-tagging-logic

Fovty opened the pull request:
## Motivation

This PR addresses the issue of automatic "latest" tagging for main branch commits, which could mistakenly promote unstable builds into production. The update ensures that the "latest" tag is only applied to stable, versioned releases. 

- Solves: #19

## Changes

- Removed automatic "latest" tagging from main branch commits.
- "Latest" tag now only applies to new release tags.
- Added logic to extract version numbers from git tags for Docker image tagging.

## TODO

- [x] I've assigned myself to this PR.

puffitos commented on .github/workflows/docker-build.yml:
```diff
@@ -15,9 +15,10 @@ jobs:
         run: |
           SHORT_SHA=$(git rev-parse --short HEAD)
-          if [[ "${{ github.ref }}" == "refs/heads/main" ]]; then
+          TAG_VERSION=$(echo ${{ github.ref }} | sed 's,refs/tags/v,,g')
```
`sed` is fine and the cool kids use it. But you know what's cooler? Using what the github API gives us. Please use github.ref_name. I also recommend taking a peek into github documentation:

- https://docs.github.com/en/actions/learn-github-actions/contexts#github-context
- https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables

Fovty replied:
Please take another good look. If github.ref_name would do string manipulations recently, I would actually have to look at the docs again

I'm not using `sed` for fun, but to strip the prefix "v" from the tag on refs/tags

puffitos replied:
Then make it even simpler. Create a separate workflow when a push with a tag occurs. 

If you're concerned about a bad actor pushing a branch name that matches semver, there should be a way to block this in github. But first would need to be a member of this repo, I guess.

Sed is used for live replacement of strings / files in Linux. Seeing it here just to grab the version from a ref name is an intuitive way to use it, but my concern is bigger. I ask you to reconcider how you approach the whole "release of a stable artifact". Use the primitives that github offers (like github.ref_type to separate between tags/branches) and eventually even use prepared github actions.

Fovty replied:
- https://github.com/caas-team/py-kube-downscaler/pull/21/commits/0a8c46e07f6f9d722c685efad1e670767c3fb908

puffitos commented on .github/workflows/docker-build.yml:
```diff
@@ -15,9 +15,10 @@ jobs:
           SHORT_SHA=$(git rev-parse --short HEAD)
-          if [[ "${{ github.ref }}" == "refs/heads/main" ]]; then
+          TAG_VERSION=$(echo ${{ github.ref }} | sed 's,refs/tags/v,,g')
+          if [[ "${{ github.ref }}" == "refs/tags/v*" ]]; then
```
Please also change this check to use the ref_name and check whether it's possible to use the "official" semver regex for the check: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string

Or use this github action to check it the tag is a valid semver: https://github.com/marketplace/actions/check-version-format-in-tag

If we use this check, the tag `velociraptor` is a valid tag and will trigger a release, with the latest tag as well. Something like that isn't likely to happen, but we should still due our due diligence here and build the check properly.

Fovty replied:
github.ref is also chosen carefully here, as the workflow runs on every push as intended.
 
An actor could name a branch v1.2.3 to trigger these actions unexpectedly since github.ref_name would just check the name without distinguishing between a branch and a tag.

- Example: `if [[ "${{ github.ref_name }}" == v* ]]` might be true for both a tag v1.2.3 and a branch named v1.2.3.

Testing the version of the tag according to semver specifications seems to me to be a valid point

puffitos replied:
Yeah, valid point. As mentioned on my other response, the github context also provides a ref_type to differentiate between branches and tags.

I still think we should separate the dev builds from the tagged release ones, as we do for example in the sparrow repo.

Fovty replied:
- https://github.com/caas-team/py-kube-downscaler/pull/21/commits/0a8c46e07f6f9d722c685efad1e670767c3fb908

puffitos commented:
Looks great. You want to adress the deprecation in this PR or would you prefer to do that in a new PR?

```
Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```

Fovty commented:
I have addressed this small fix right here: https://github.com/caas-team/py-kube-downscaler/pull/21/commits/4192efa122a4aac9f263388f2761282be8fe305b and fixed the new introduced bug here: https://github.com/caas-team/py-kube-downscaler/pull/21/commits/3031889212fc89a0354611400df8ef8e0f4cb59b. Now it should be fine

> Looks great. You want to adress the deprecation in this PR or would you prefer to do that in a new PR?
> 
> ```
> Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
> ```
